File: bindings/java/src/main/java/com/moofile/Filters.java
package com.moofile;

import java.util.ArrayList;
import java.util.List;

/**
 * Static factories for MooFile's MongoDB-style filters.
 *
 * <p>Use static imports for concise, idiomatic Java queries:
 *
 * <pre>{@code
 * import static com.moofile.Filters.*;
 *
 * Document activeAdults = and(
 *     gte("age", 18),
 *     eq("status", "active"));
 *
 * List<Document> people = db.find(activeAdults);
 * }</pre>
 *
 * <p>These factories cover the filter operators MooFile supports. They return
 * ordinary {@link Document} instances, so they work with every existing API
 * that accepts a filter.
 */
public final class Filters {
    private Filters() { }

    /** Match a field exactly. */
    public static Document eq(String field, Object value) {
        return Document.of(requireField(field), value);
    }

    /** Match a field that does not equal a value. */
    public static Document ne(String field, Object value) {
        return fieldOperator(field, "$ne", value);
    }

    public static Document gt(String field, Object value)  { return fieldOperator(field, "$gt", value); }
    public static Document gte(String field, Object value) { return fieldOperator(field, "$gte", value); }
    public static Document lt(String field, Object value)  { return fieldOperator(field, "$lt", value); }
    public static Document lte(String field, Object value) { return fieldOperator(field, "$lte", value); }

    /** Match a field equal to any item in {@code values}. */
    public static Document in(String field, Iterable<?> values) {
        return fieldOperator(field, "$in", valuesToList(values));
    }

    /** Match a field unequal to every item in {@code values}. */
    public static Document nin(String field, Iterable<?> values) {
        return fieldOperator(field, "$nin", valuesToList(values));
    }

    /** Match documents where a field is present and non-null. */
    public static Document exists(String field) {
        return exists(field, true);
    }

    /** Match documents where a field is present/non-null when {@code exists} is true. */
    public static Document exists(String field, boolean exists) {
        return fieldOperator(field, "$exists", exists);
    }

    /** Match an array with at least one element satisfying {@code filter}. */
    public static Document elemMatch(String field, Document filter) {
        if (filter == null) throw new IllegalArgumentException("filter must not be null");
        return fieldOperator(field, "$elemMatch", filter);
    }

    /** Match documents satisfying every supplied filter. */
    public static Document and(Document... filters) {
        return logical("$and", filters);
    }

    /** Match documents satisfying at least one supplied filter. */
    public static Document or(Document... filters) {
        return logical("$or", filters);
    }

    /** Invert a filter. */
    public static Document not(Document filter) {
        if (filter == null) throw new IllegalArgumentException("filter must not be null");
        return Document.of("$not", filter);
    }

    private static Document fieldOperator(String field, String operator, Object value) {
        return Document.of(requireField(field), Document.of(operator, value));
    }

    private static Document logical(String operator, Document... filters) {
        if (filters == null) throw new IllegalArgumentException("filters must not be null");
        List<Object> documents = new ArrayList<>(filters.length);
        for (Document filter : filters) {
            if (filter == null) throw new IllegalArgumentException("filters must not contain null");
            documents.add(filter);
        }
        return Document.of(operator, documents);
    }

    private static List<Object> valuesToList(Iterable<?> values) {
        if (values == null) throw new IllegalArgumentException("values must not be null");
        List<Object> out = new ArrayList<>();
        for (Object value : values) out.add(value);
        return out;
    }

    private static String requireField(String field) {
        if (field == null || field.isEmpty()) {
            throw new IllegalArgumentException("field must not be null or empty");
        }
        return field;
    }
}

File: bindings/csharp/Moofile/Filters.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

namespace Moofile;

/// <summary>
/// A strongly typed MooFile filter. Build one with <see cref="Builders{TDocument}"/>.
/// </summary>
/// <typeparam name="TDocument">
/// A CLR type whose public property names correspond to fields in a MooFile document.
/// </typeparam>
public sealed class FilterDefinition<TDocument>
{
    private readonly Document _document;

    internal FilterDefinition(Document document) => _document = document;

    /// <summary>
    /// Return the MongoDB-style document sent to MooFile. This is useful when
    /// interoperating with APIs that accept an untyped <see cref="Document"/>.
    /// </summary>
    public Document ToDocument() => _document;
}

/// <summary>Entry point for strongly typed filter builders.</summary>
/// <typeparam name="TDocument">The CLR type used to select field names.</typeparam>
/// <remarks>
/// This is an expression-based convenience API over MooFile's supported filter
/// operators; it is not a general LINQ provider. Selectors must be direct,
/// top-level properties, such as <c>person =&gt; person.Age</c>.
/// </remarks>
public static class Builders<TDocument>
{
    /// <summary>Build filters for <typeparamref name="TDocument"/>.</summary>
    public static FilterDefinitionBuilder<TDocument> Filter { get; } = new();
}

/// <summary>Build MongoDB-style filters using strongly typed property selectors.</summary>
public sealed class FilterDefinitionBuilder<TDocument>
{
    /// <summary>Match a field exactly.</summary>
    public FilterDefinition<TDocument> Eq<TField>(Expression<Func<TDocument, TField>> field, TField value) =>
        Field(field, value);

    /// <summary>Match a field that does not equal a value.</summary>
    public FilterDefinition<TDocument> Ne<TField>(Expression<Func<TDocument, TField>> field, TField value) =>
        FieldOperator(field, "$ne", value);

    public FilterDefinition<TDocument> Gt<TField>(Expression<Func<TDocument, TField>> field, TField value) =>
        FieldOperator(field, "$gt", value);

    public FilterDefinition<TDocument> Gte<TField>(Expression<Func<TDocument, TField>> field, TField value) =>
        FieldOperator(field, "$gte", value);

    public FilterDefinition<TDocument> Lt<TField>(Expression<Func<TDocument, TField>> field, TField value) =>
        FieldOperator(field, "$lt", value);

    public FilterDefinition<TDocument> Lte<TField>(Expression<Func<TDocument, TField>> field, TField value) =>
        FieldOperator(field, "$lte", value);

    public FilterDefinition<TDocument> In<TField>(Expression<Func<TDocument, TField>> field, IEnumerable<TField> values) =>
        FieldOperator(field, "$in", values);

    public FilterDefinition<TDocument> Nin<TField>(Expression<Func<TDocument, TField>> field, IEnumerable<TField> values) =>
        FieldOperator(field, "$nin", values);

    /// <summary>Match documents where a field is present (or absent).</summary>
    public FilterDefinition<TDocument> Exists<TField>(Expression<Func<TDocument, TField>> field, bool exists = true) =>
        FieldOperator(field, "$exists", exists);

    /// <summary>Match an array containing an element that satisfies <paramref name="filter"/>.</summary>
    public FilterDefinition<TDocument> ElemMatch<TCollection, TItem>(
        Expression<Func<TDocument, TCollection>> field,
        FilterDefinition<TItem> filter)
        where TCollection : IEnumerable<TItem> =>
        FieldOperator(field, "$elemMatch", filter?.ToDocument() ?? throw new ArgumentNullException(nameof(filter)));

    /// <summary>Match documents satisfying every supplied filter.</summary>
    public FilterDefinition<TDocument> And(params FilterDefinition<TDocument>[] filters) =>
        Logical("$and", filters);

    /// <summary>Match documents satisfying at least one supplied filter.</summary>
    public FilterDefinition<TDocument> Or(params FilterDefinition<TDocument>[] filters) =>
        Logical("$or", filters);

    /// <summary>Invert a filter.</summary>
    public FilterDefinition<TDocument> Not(FilterDefinition<TDocument> filter)
    {
        ArgumentNullException.ThrowIfNull(filter);
        return new FilterDefinition<TDocument>(Document.Of("$not", filter.ToDocument()));
    }

    private static FilterDefinition<TDocument> Field<TField>(
        Expression<Func<TDocument, TField>> field, TField value) =>
        new(Document.Of(FieldName(field), value));

    private static FilterDefinition<TDocument> FieldOperator<TField>(
        Expression<Func<TDocument, TField>> field, string op, object? value) =>
        new(Document.Of(FieldName(field), Document.Of(op, value)));

    private static FilterDefinition<TDocument> Logical(string op, IEnumerable<FilterDefinition<TDocument>> filters)
    {
        ArgumentNullException.ThrowIfNull(filters);
        var docs = filters.Select(filter =>
            filter?.ToDocument() ?? throw new ArgumentException("filters cannot contain null", nameof(filters)))
            .Cast<object?>().ToList();
        return new FilterDefinition<TDocument>(Document.Of(op, docs));
    }

    private static string FieldName<TField>(Expression<Func<TDocument, TField>> selector)
    {
        ArgumentNullException.ThrowIfNull(selector);
        Expression body = selector.Body;
        while (body is UnaryExpression { NodeType: ExpressionType.Convert or ExpressionType.ConvertChecked } convert)
            body = convert.Operand;

        if (body is not MemberExpression { Expression: ParameterExpression } member)
        {
            throw new ArgumentException(
                "Field selectors must be direct top-level properties, for example person => person.Age.",
                nameof(selector));
        }
        return member.Member.Name;
    }
}
